### PR TITLE
feat: send message api

### PR DESCRIPTION
**Pull request by mhuynh5757**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [VBLOCKS-1075](https://issues.corp.twilio.com/browse/VBLOCKS-1075)

### Description

This PR adds the send message API for SDK Eventing.

### Before review
* [ ] Updated CHANGELOG.md if necessary
* [ ] Added unit tests if necessary
* [ ] Updated affected documentation
* [ ] Verified locally with `npm test`
* [ ] Manually sanity tested running locally
* [ ] Ready for review

### Before merge
* [ ] Got one or more +1s
* [ ] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

File: lib/twilio/call.ts [hunk 1]
```diff
@@ -851,6 +862,45 @@ class Call extends EventEmitter {
     }
   }
 
+  /**
+   * Send a message to Twilio.
```
Comment (charliesantos) [inline lib/twilio/call.ts:866]: Need to expand more on this. Just calling it out. No need to address now. We'll do it as part of polishing up the docs.

File: lib/twilio/call.ts [hunk 2]
```diff
@@ -851,6 +862,45 @@ class Call extends EventEmitter {
...
+   * call.sendMessage(Call.MessageType.UserDefinedMessage, {
+   *   ahoy: 'world!,
+   * });
+   * ```
+   *
+   * @param messageType - The type of the message to send to Twilio.
+   * @param content - The content of the message to send to Twilio.
+   */
+  sendMessage(messageType: Call.MessageType, content: any) {
+    if (typeof messageType !== 'string') {
```
Comment (charliesantos) [inline lib/twilio/call.ts:879]: Should we check for empty strings?

File: lib/twilio/call.ts [hunk 3]
```diff
@@ -851,6 +862,45 @@ class Call extends EventEmitter {
...
+   * ```ts
+   * call.sendMessage(Call.MessageType.UserDefinedMessage, {
+   *   ahoy: 'world!,
+   * });
+   * ```
+   *
+   * @param messageType - The type of the message to send to Twilio.
+   * @param content - The content of the message to send to Twilio.
+   */
+  sendMessage(messageType: Call.MessageType, content: any) {
```
Comment (charliesantos) [inline lib/twilio/call.ts:878]: Should this be a promise? Or is that a separate task? Right now it just a straight hand off to signaling.

File: lib/twilio/call.ts [hunk 4]
```diff
@@ -851,6 +862,45 @@ class Call extends EventEmitter {
...
+        'Could not send CallMessage; Signaling channel is disconnected',
+      );
+    }
+
+    const callSid = this.parameters.CallSid;
+    if (typeof callSid !== 'string') {
+      throw new InvalidStateError(
+        'Could not send CallMessage; Call has no CallSid',
+      );
+    }
```
Comment (charliesantos) [inline lib/twilio/call.ts:897]: I think it's better to check call status so you don't need these 2 checks. You will also probably cover more scenario. Something like
```ts
if (this._status !== Call.State.Open) {
  throw new InvalidStateError(
      `Could not send message. Call status is ${this._status}`
  );
}
``

File: lib/twilio/pstream.js [hunk 5]
```diff
@@ -177,6 +177,22 @@ PStream.prototype.setToken = function(token) {
   this._publish('listen', payload);
 };
 
+PStream.prototype.sendMessage = function(
```
Comment (charliesantos) [inline lib/twilio/pstream.js:180]: Something to keep at the back of our heads. This pattern is old school and needs rework.

File: package.json [hunk 6]
```diff
@@ -122,8 +122,10 @@
   "dependencies": {
     "@twilio/audioplayer": "1.0.6",
     "@twilio/voice-errors": "1.1.1",
+    "@types/md5": "^2.3.2",
     "backoff": "2.5.0",
     "loglevel": "1.6.7",
+    "md5": "^2.3.0",
```
Comment (charliesantos) [inline package.json:128]: Let's use exact version for consistency. I know we now have a lock file, but we should keep this consistent.

Comment (charliesantos) [reply package.json:128]: What is the size difference?

File: lib/twilio/uuid.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import * as md5 from 'md5';
+import { NotSupportedError } from '../twilio/errors';
+
+export function generateUuid(): string {
```
Comment (charliesantos) [inline lib/twilio/uuid.ts:4]: Why are we exporting it? For unit tests? If so, we should still keep it private. We can hit this function by testing generateVoiceEventSid instead.

**Review by charliesantos — CHANGES_REQUESTED** [review]

File: lib/twilio/call.ts (hunk 2 above)
Comment (mhuynh5757) [reply lib/twilio/call.ts:879]: Good point, I will add that.

File: lib/twilio/call.ts (hunk 3 above)
Comment (mhuynh5757) [reply lib/twilio/call.ts:878]: Yeah so that's the task that's rolling over, using the `ack` from signaling.

File: lib/twilio/call.ts (hunk 4 above)
Comment (mhuynh5757) [reply lib/twilio/call.ts:897]: Sounds good, will do.

File: package.json (hunk 6 above)
Comment (mhuynh5757) [reply package.json:128]: What's the best way to calculate that? Just look at the size of the two new packages?

File: lib/twilio/uuid.ts (hunk 7 above)
Comment (mhuynh5757) [reply lib/twilio/uuid.ts:4]: Hmm, I can see us using in other things down the line maybe, but we can export it then. I'll make it non-exported.

File: package.json (hunk 6 above)
Comment (charliesantos) [reply package.json:128]: Build the package, check the size of the minified/unminified file and compare it to the previous release (download from tags or npm)

File: lib/twilio/uuid.ts (hunk 7 above)
Comment (charliesantos) [reply lib/twilio/uuid.ts:4]: Right. We should not export anything that is not needed. Only do it when you need it. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) 😉

File: lib/twilio/call.ts (hunk 4 above)
Comment (mhuynh5757) [reply lib/twilio/call.ts:897]: Do you think it's still worth type-guarding `this.parameters.CallSid`?

Comment (charliesantos) [reply lib/twilio/call.ts:897]: Do you mean keep the `typeof callSid !== 'string'` check? I don't think we need it. Check the status that it is open should cover that.

File: package.json (hunk 6 above)
Comment (mhuynh5757) [reply package.json:128]: `md5` and `@types/md5` in the end adds 10kb; `twilio.min.js` goes from 335kb to 345kb

File: lib/twilio/call.ts (hunk 4 above)
Comment (charliesantos) [reply lib/twilio/call.ts]: @mhuynh5757 I just learned base on this [discussion](https://twilio.slack.com/archives/C02U2QQCMME/p1663964471252479) that client should be able to receive events outside of "open" status, such as ringing. This means, we should only check here if there's a websocket connection or not.

Comment (mhuynh5757) [reply lib/twilio/call.ts]: What about a CallSid? This is sending not receiving right? Seems kind of weird to want to send a message during ringing.

Comment (charliesantos) [reply lib/twilio/call.ts]: It's a valid use case. The customer can use it to send status events to their backend or other clients.

Comment (mhuynh5757) [reply lib/twilio/call.ts]: Ok, sounds solid then.

Comment (mhuynh5757) [reply lib/twilio/call.ts]: Do we want to add the CallSid check back in?

Comment (charliesantos) [reply lib/twilio/call.ts]: No, we don't want to check for callsid. I don't think callsid is generated until the call is established no?

Comment (mhuynh5757) [reply lib/twilio/call.ts]: Yes, but we always send the `CallSid` with every `message` package over VSP.

**Comment by charliesantos** [comment]
Latest changes looks good. Please address the callsid related discussion then we can merge
https://twilio.slack.com/archives/C042FCK9WSD/p1663968312559099?thread_ts=1663961297.212249&cid=C042FCK9WSD

**Review by charliesantos — APPROVED** [review]

**Comment by charliesantos** [comment]
I'll go ahead and merge this one @mhuynh5757

**Pull request merged**